rfcs/0001-messages.md

```diff
@@ -0,0 +1,260 @@
+
+```rust
+msg.header.validate();
+```
+
+## Serialization/Deserialization
+
+Each message struct needs to serialize to bytes representation before it is sent through the wire and must deserialize to the same struct (round trip) on the receiver side. We use `serde` and macro derivations (`Serialize` and `Deserialize`) to automatically implement where possible.
+
+This will require deriving serde in several types defined in `frost.rs`. 
+Manual implementation of serialization/deserialization will be located at a new mod `src/frost/serialize.rs`.
+
+### Header
+
+The `Header` part of the message is 4 bytes total:
+
+Bytes | Field name | Data type
+------|------------|-----------
+1     | msg_type   | u8
+1     | version    | u8
+1     | sender     | u8
+1     | receiver   | u8
+
+### Primitive types
+
+`Payload`s use data types that we need to specify first. We have 3 primitive types inside the payload messages:
+
+#### `Scalar`
+
+`Scalar` is a better name for `jubjub::Fr` and this is a `[u64; 4]` as documented in https://github.com/zkcrypto/jubjub/blob/main/src/fr.rs#L16
+
+#### `Commitment`
+
+`Commitment` is a wrapper of `jubjub::ExtendedPoint` and this is a structure with 5 `jubjub::Fq`s as defined in https://github.com/zkcrypto/jubjub/blob/main/src/lib.rs#L128-L134
+
+Each `Fq` needed to form a `jubjub::ExtendedPoint` are `Scalar`s of `bls12_381` crate. Scalar here is `[u64; 4]` as documented in https://github.com/zkcrypto/bls12_381/blob/main/src/scalar.rs#L16
+
+#### `ExtendedPoint`
+
+`ExtendedPoint` was detailed above, it is 5 `[u64; 4]`. The total size of an `ExtendedPoint` is 1280 bytes.
+
+### FROST types
+
+`Payload`s also use some types that are defined in the `redjubjub` crate. Here we describe them from a serialization point of view.
+
+#### `VerificationKey<SpendAuth>`
+
+Defined in `verification_key.rs` it consist of 1 `ExtendedPoint` and 1 `VerificationKeyBytes` which is also defined in the same file and consist of 1 `[u8; 32]`.
+
+#### `Signature<SpendAuth>`
+
+Defined in `signature.rs` consist of 2 `[u8; 32]` arrays.
+
+### Payload
+
+Payload part of the message is variable in size and depends on message type.
+
+#### `MsgDealerBroadcast`
+
+Bytes  | Field name  | Data type
+-------|-------------|-----------
+256    | secret_key  | Scalar
+1280*n | commitments | [Commitment; n]
+1280+32| group_public| VerificationKey<SpendAuth>
+
+#### `MsgCommitments`
+
+Bytes | Field name | Data type
+------|------------|-----------
+1280  | hiding     | ExtendedPoint
+1280  | binding    | ExtendedPoint
+
+#### `MsgSigningPackage`
+
+Bytes      | Field name     | Data type
+-----------|----------------|-----------
+1+(1280*n) | signing_package| u8 [Commitment; n]
+
+#### `SignatureShare`
+
+Bytes | Field name | Data type
+------|------------|-----------
+256   | signature  | Scalar
+
+#### `MsgFinalSignature`
+
+Bytes | Field name | Data type
+------|------------|-----------
+64    | signature  | Signature<SpendAuth>
```

Comment:
What is the byte order of all these multi-byte fields?
https://en.wikipedia.org/wiki/Endianness

Is the byte order applied to their components, or to the entire field?
For example, for some byte orders, the serialization of a `[u64; 4]` and a `[u256; 1]` are different, even though they both have 256 bytes.

Reply:
@dconnolly might be able to help you here.

Reply:
Everything will be in little-endian order unless specified otherwise. I added a note about this https://github.com/ZcashFoundation/redjubjub/pull/85/commits/1ee805987c24d83231a5f20aa8313d09a1ccf48f

Reply:
Is the byte order applied to each sub-component, or to the entire field?

For example, for some byte orders, the serialization of a [u64; 4] and a [u256; 1] are different, even though they both have 256 bytes.

Reply:
In general, serialization should work with `[u8; N]` arrays, and then describe how those arrays are turned into big integers. (For example: `[u8; 256]` becomes a little-Endi an `u256`.)

`[u64; 4]` is an arithmetic implementation detail, so it shouldn't appear in the serialization spec.

Reply:
This is obsoleted by https://github.com/ZcashFoundation/redjubjub/pull/85#discussion_r629008634